### PR TITLE
[rtl872x] correctly handle SecureFault

### PR DESCRIPTION
**Pull request by avtolstoy**
See [SC121377]

### CODE REVIEW

**Review by mrlambchop — APPROVED** [review]

File: bootloader/src/rtl872x/rtl_it.c [hunk 1]
```diff
@@ -128,6 +143,30 @@ void UsageFault_Handler(void) {
...
+    Fault_Handler(SecureFault);
+}
+
+__attribute__((always_inline)) inline
+void jump_to_nonsecure(u32 Addr) {
+    __ASM volatile ("MOV r0, %0\n\t"
+        "BLXNS   r0\n\t" : : "r" (Addr));
+}
+
+__attribute__((used, section(".secure.ram.text")))
```
Comment (YutingYou) [inline bootloader/src/rtl872x/rtl_it.c:156]: This line appears the key modification, it works for my test app as well.

@avtolstoy I was wondering why running the SecureFault_Handler in the secure RAM is not enough, we have to put it in the `.secure.ram.text`?

File: bootloader/src/rtl872x/rtl_it.c [hunk 2]
```diff
@@ -128,6 +143,30 @@ void UsageFault_Handler(void) {
     Fault_Handler(UsageFault);
 }
 
+void SecureFault_Handler_NS(void) {
```
Comment (YutingYou) [inline bootloader/src/rtl872x/rtl_it.c:146]: As per the map file and IDAU configuration `10006318` seems to reside in secure region.
```
10006318 <SecureFault_Handler_NS>:
    /* Go to infinite loop when Usage Fault exception occurs */
    Fault_Handler(UsageFault);
}
```

```
const TZ_CFG_TypeDef tz_config[]= {
    //  Start                    End                NSC
    {0x40000000,            0x50000000-1,            0},  // entry0: Peripherals NS
    {0x1010A000,            0x101D4000-1,            0},  // entry1: IROM & DROM NS
    {0x100E0000,            0x10100000-1,            0},  // entry2: BT/WIFI Extention SRAM
    {0x00000000,            0x08060000-1,            0},  // entry3: KM0 ROM, KM0 SRAM, Retention SRAM, PSRAM, KM0 FLASH
    {0x08060000,            0x1007C000-1,            0},  // entry4: KM4 Flash, KM4 SRAM
    {0xFFFFFFFF,            0xFFFFFFFF,              0},  // entry5: None
    {0xFFFFFFFF,            0xFFFFFFFF,              0},  // entry6: None
    {0xFFFFFFFF,            0xFFFFFFFF,              0},  // entry7: None
};
```
Screenshot 2023-09-22 at 11 39 36

Comment (avtolstoy) [reply bootloader/src/rtl872x/rtl_it.c:146]: It applies to `{0x08060000,            0x1007C000-1,            0},  // entry4: KM4 Flash, KM4 SRAM` entry, which is marked as NS by IDAU.

Comment (YutingYou) [reply bootloader/src/rtl872x/rtl_it.c:146]: Oh, I missed that.

**Review by YutingYou — APPROVED** [review]

**Comment by YutingYou** [comment]
Tested with below app, secure fault handler in VTOR_NS can be handled.
```
void setup() {
    uint32_t *test_ptr = (uint32_t *)0x1007EFFC;    // SecureFault
    // uint32_t *test_ptr = (uint32_t *)0x0;        // BusFault
    volatile uint32_t test_value = 0;
    test_value = *test_ptr;
}
```

**Pull request merged**